### PR TITLE
Remove some `@ts-expect-error`

### LINKED ISSUE #337: Fix TypeScript ignore directives
We have a lot of ignore directives now. We should try to fix the types:

https://github.com/sindresorhus/ky/search?q=%22@ts-expect-error%22&type=

### PR DESCRIPTION
**Pull request by Beraliv**
- [x] fixed in production code
- [x] fixed in tests
- [x] ts:check errors
- [x] tests
- [x] resolved conflicts

But I didn't remove every `@ts-expect-error` as sometimes it's needed (for error validation in test cases)

Fixes #337

### CODE REVIEW

File: source/types/hooks.ts [hunk 1]
```diff
@@ -8,7 +8,6 @@ export type BeforeRequestHook = (
 
 export type BeforeRetryState = {
 	request: Request;
-	response: Response;
```
Comment (Beraliv) [inline source/types/hooks.ts:11]: I'm not sure if response is needed

In docs it's not mentioned, in tests I didn't find a usage too

File: source/index.ts [hunk 2]
```diff
@@ -6,14 +6,17 @@ import type {ky as KyInterface} from './types/ky.js';
...
 import {validateAndMerge} from './utils/merge.js';
 
+type Mutable<T> = {
+	-readonly[P in keyof T]: T[P]
+};
+
 const createInstance = (defaults?: Partial<Options>): KyInterface => {
 	// eslint-disable-next-line @typescript-eslint/promise-function-async
-	const ky = (input: Input, options?: Options) => Ky.create(input, validateAndMerge(defaults, options));
+	const ky: Partial<Mutable<KyInterface>> = (input: Input, options?: Options) => Ky.create(input, validateAndMerge(defaults, options));
```
Comment (Beraliv) [inline source/index.ts:15]: So inside we declare it mutable and optional

So we can assign it later

File: source/index.ts [hunk 3]
```diff
@@ -6,14 +6,17 @@ import type {ky as KyInterface} from './types/ky.js';
 import type {Input, Options} from './types/options.js';
 import {validateAndMerge} from './utils/merge.js';
 
+type Mutable<T> = {
+	-readonly[P in keyof T]: T[P]
+};
```
Comment (Beraliv) [inline source/index.ts:11]: It's needed for `KyInterface` as it has `readonly stop` which needs to make mutable below (to be able to assign the value)

File: source/core/constants.ts [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
 export const supportsAbortController = typeof globalThis.AbortController === 'function';
 export const supportsStreams = typeof globalThis.ReadableStream === 'function';
 export const supportsFormData = typeof globalThis.FormData === 'function';
 
-export const requestMethods: readonly HTTPMethod[] = ['get', 'post', 'put', 'patch', 'head', 'delete'];
+export const requestMethods = ['get', 'post', 'put', 'patch', 'head', 'delete'] as const;
+
+type cases = [
+	Expect<Equal<typeof requestMethods[number], HTTPMethod>>,
+]
```
Comment (Beraliv) [inline source/core/constants.ts:12]: instead of declaring `requestMethods` as `readonly HTTPMethod[]`, you make it readonly as is

and then check here in `cases` that values are exactly the same

File: source/core/constants.ts [hunk 5]
```diff
@@ -1,18 +1,23 @@
...
 
 export const responseTypes = {
 	json: 'application/json',
 	text: 'text/*',
 	formData: 'multipart/form-data',
 	arrayBuffer: '*/*',
 	blob: '*/*'
-};
+} as const;
```
Comment (Beraliv) [inline source/core/constants.ts:20]: this is needed to infer key and values for `Object.entries` later

File: source/core/Ky.ts [hunk 6]
```diff
@@ -145,10 +149,9 @@ export class Ky {
 		};
 
 		const isRetriableMethod = ky._options.retry.methods.includes(ky.request.method.toLowerCase());
-		const result = isRetriableMethod ? ky._retry(fn) : fn();
+		const result = (isRetriableMethod ? ky._retry(fn) : fn()) as ResponsePromise;
 
-		for (const [type, mimeType] of Object.entries(responseTypes)) {
-			// @ts-expect-error not sure how to properly type this!
+		for (const [type, mimeType] of Object.entries(responseTypes) as ObjectEntries<typeof responseTypes>) {
```
Comment (Beraliv) [inline source/core/Ky.ts:154]: the problem here is that `Object.entries` has incorrect `ReturnType`

So what I do:

1. I created `ObjectEntries` helper to infer it (however, it's possible to redefine it globally)
2. `result` as `ResponsePromise` now has all the methods for `type` so we can call it without issues

File: package.json [hunk 7]
```diff
@@ -50,6 +50,7 @@
 	],
 	"devDependencies": {
 		"@sindresorhus/tsconfig": "^1.0.1",
+		"@type-challenges/utils": "^0.1.1",
```
Comment (Beraliv) [inline package.json:53]: is needed for `Expect` and `Equal`

however, I can define it without dependency

File: test/browser.ts [hunk 8]
```diff
@@ -157,9 +157,10 @@ test(
 				};
 			});
 		}, server.url);
-		// @ts-expect-error
+		if (typeof error !== 'object') {
+			throw new Error('Expected to have an object error');
+		}
```
Comment (Beraliv) [inline test/browser.ts:162]: it's validated with `throw` so it's narrowed the `object` now

File: test/fetch.ts [hunk 9]
```diff
@@ -4,9 +4,11 @@ import ky from '../source/index.js';
 test.serial('relative URLs are passed to fetch unresolved', async t => {
 	const originalFetch = globalThis.fetch;
 	globalThis.fetch = async input => {
-		// @ts-expect-error @TODO
+		if (typeof input !== 'object') {
+			throw new Error('Expect to have an object request');
+		}
```
Comment (Beraliv) [inline test/fetch.ts:9]: it's validated with `throw` so it's narrowed the `object` now

File: source/types/options.ts [hunk 10]
```diff
@@ -71,7 +71,7 @@ export interface Options extends Omit<RequestInit, 'headers'> {
 
 	Accepts any plain object or value, which will be `JSON.stringify()`'d and sent in the body with the correct header set.
 	*/
-	json?: unknown;
+	json?: Record<string, unknown>;
```
Comment (Beraliv) [inline source/types/options.ts:74]: Correct me if it can be not only an object

File: test/hooks.ts [hunk 11]
```diff
@@ -270,19 +271,19 @@ test('`afterResponse` hook is called with request, normalized options, and respo
...
-						// @ts-expect-error
 						async (request, options, response) => {
 							if (response.status === 403) {
 								// Retry request with valid token
 								return ky(request, {
 									...options,
 									json: {
-										// @ts-expect-error @TODO
-										...options.json,
+										...(options as Options).json,
```
Comment (Beraliv) [inline test/hooks.ts:280]: we cannot use spread here if it's not an object

if it's valid only in test, let's cast it here

File: test/hooks.ts [hunk 12]
```diff
@@ -270,19 +271,19 @@ test('`afterResponse` hook is called with request, normalized options, and respo
...
 										token: 'valid:token'
 									}
 								});
 							}
+
+							return undefined;
```
Comment (Beraliv) [inline test/hooks.ts:286]: should return something in other branches

File: test/http-error.ts [hunk 13]
```diff
@@ -1,18 +1,18 @@
...
+type Mutable<T> = {
+	-readonly[P in keyof T]: T[P]
+};
+
 function createFakeResponse({status, statusText}: {status?: number; statusText?: string}): Response {
 	// Start with a realistic fetch Response.
-	const response = {...new Response()};
+	const response: Partial<Mutable<Response>> = {...new Response()};
 
-	// Assign these values because the Response() constructor doesn't
```
Comment (Beraliv) [inline test/http-error.ts:8]: There are 2 problems here:

1. it is `readonly`
2. it doesn't accept `undefined`

File: test/http-error.ts [hunk 14]
```diff
@@ -1,18 +1,18 @@
 import test from 'ava';
 import {HTTPError} from '../source/index.js';
 
+type Mutable<T> = {
```
Comment (Beraliv) [inline test/http-error.ts:4]: Let me know if I can move it to some place to reuse it (was used twice in the codebase)

**Comment by Beraliv** [comment]
At the moment cannot run tests

Can you help me here @sindresorhus ?

File: source/core/Ky.ts [hunk 15]
```diff
@@ -145,14 +149,14 @@ export class Ky {
 		};
 
 		const isRetriableMethod = ky._options.retry.methods.includes(ky.request.method.toLowerCase());
-		const result = isRetriableMethod ? ky._retry(fn) : fn();
+		const result = (isRetriableMethod ? ky._retry(fn) : fn()) as ResponsePromise;
```
Comment (sindresorhus) [inline source/core/Ky.ts:152]: Sidenote: We should also try to reduce use of `as` as it just overrides the type no matter what.

File: source/types/options.ts (hunk 10 above)
Comment (sindresorhus) [reply source/types/options.ts:74]: JSON can have top-level array or primitive value.

File: test/hooks.ts (hunk 11 above)
Comment (sindresorhus) [reply test/hooks.ts:280]: The real issue is that the `options` parameter doesn't have the correct type.

File: test/http-error.ts (hunk 14 above)
Comment (sindresorhus) [reply test/http-error.ts:4]: You can move the type utilities to a `types` file in https://github.com/sindresorhus/ky/tree/main/source/utils

File: source/core/Ky.ts [hunk 16]
```diff
@@ -145,14 +149,14 @@ export class Ky {
...
 
-		for (const [type, mimeType] of Object.entries(responseTypes)) {
-			// @ts-expect-error not sure how to properly type this!
+		for (const [type, mimeType] of Object.entries(responseTypes) as ObjectEntries<typeof responseTypes>) {
 			result[type] = async () => {
 				// eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
 				ky.request.headers.set('accept', ky.request.headers.get('accept') || mimeType);
 
+				// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion, @typescript-eslint/non-nullable-type-assertion-style
```
Comment (sindresorhus) [inline source/core/Ky.ts:159]: Why are you silencing the listing? If it's valid, it would be useful with a code comment about why.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/core/Ky.ts (hunk 16 above)
Comment (Beraliv) [reply source/core/Ky.ts:159]: No idea here, somehow `npm run test` throws an error here

File: source/core/Ky.ts (hunk 15 above)
Comment (Beraliv) [reply source/core/Ky.ts:152]: You anyway do that below, so I moved it here

Is it alright?

File: test/hooks.ts (hunk 11 above)
Comment (Beraliv) [reply test/hooks.ts:280]: What is the expected type here?

File: source/types/options.ts (hunk 10 above)
Comment (Beraliv) [reply source/types/options.ts:74]: So would it be ok to have `PrimitiveType | Record<string, unknown>`?

```ts
type PrimitiveType = number | string | boolean | undefined | null;
```

File: source/core/Ky.ts (hunk 15 above)
Comment (sindresorhus) [reply source/core/Ky.ts:152]: I mean didn't mean to apply that you added it. Was just saying in general, we should try to reduce the `as` whenever possible as they're often as bad as `@ts-expect-error`.

File: test/http-error.ts (hunk 14 above)
Comment (Beraliv) [reply test/http-error.ts]: Moved to source/utils ✅

File: source/utils/Mutable.ts [hunk 17]
```diff
@@ -0,0 +1,3 @@
+export type Mutable<T> = {
```
Comment (sindresorhus) [inline source/utils/Mutable.ts:1]: The file should be named `types.ts`. I don't want a separate file for each simple type utility.

Comment (Beraliv) [reply source/utils/Mutable.ts]: yeah, that's working for me

Done ✅

File: source/core/Ky.ts (hunk 15 above)
Comment (Beraliv) [reply source/core/Ky.ts:152]: Okay, I understand that 👍

File: test/hooks.ts (hunk 11 above)
Comment (Beraliv) [reply test/hooks.ts:280]: Also, should it be connected to the `json`  we have in test?

In this case we need to have Generic type:

It will be `Options<Json extends BasicJson = BasicJson>` instead of `Options`

What do you think?

File: source/types/options.ts (hunk 10 above)
Comment (Beraliv) [reply source/types/options.ts:74]: Btw, what is the expected value after calling `.json()` if you pass primitives?

Here in test case I see that it's string instead of `boolean`:

```ts
test('POST JSON with falsey value', async t => {
	// #222
	const server = await createHttpTestServer({bodyParser: false});
	server.post('/', async (request, response) => {
		response.json(await parseRawBody(request));
	});

	const json = false;
	const responseJson = await ky.post(server.url, {json}).json();

	t.deepEqual(responseJson, json.toString());

	await server.close();
});
```

Can you please describe the rules so I will be on the same page?

File: test/hooks.ts (hunk 11 above)
Comment (Beraliv) [reply test/hooks.ts:280]: @sindresorhus I can suggest these changes regarding more stricter mapping of `Json` and the types in hooks and `json`

You can have a look at the changes (I didn't finish it, but you can already see the possible way) – https://github.com/Beraliv/ky/pull/1

wdyt?

File: source/types/options.ts (hunk 10 above)
Comment (sindresorhus) [reply source/types/options.ts:74]: > So would it be ok to have Primitive | Record<string, unknown>?

No, it used to be that, but it caused problems with `interface`s. `git blame` it and you'll find the relevant thread.

File: test/hooks.ts (hunk 11 above)
Comment (sindresorhus) [reply test/hooks.ts:280]: I think `...options.json as Record<string, unknown>` would be fine here. We know it's an object, so it's fine to force cast.

File: test/hooks.ts [hunk 18]
```diff
@@ -340,8 +341,7 @@ test('beforeRetry hook is never called for the initial request', async t => {
 				hooks: {
 					beforeRetry: [
 						({options}) => {
-							// @ts-expect-error @TODO
-							options.headers!.set('unicorn', fixture);
+							(options.headers as Partial<Headers> | undefined)?.set?.('unicorn', fixture);
```
Comment (sindresorhus) [inline test/hooks.ts:344]: Why is this needed? Is the `options.headers` type not correct?

File: test/hooks.ts (hunk 11 above)
Comment (Beraliv) [reply test/hooks.ts:280]: Do you think https://github.com/Beraliv/ky/pull/1 will be too much?

**Comment by Beraliv** [comment]
@sindresorhus I go back to this MR at the end of this week

File: test/hooks.ts (hunk 11 above)
Comment (sindresorhus) [reply test/hooks.ts:280]: I think:

> Instead of ky.get(...).json<IncomingHttpHeaders>() we will have ky.get<IncomingHttpHeaders>(...).json()

is a good idea, but making it extend a JSON type is problematic for many reasons. TS is just not good enough to make it workable. https://github.com/sindresorhus/ky/pull/133#discussion_r282433150

File: test/hooks.ts (hunk 18 above)
Comment (Beraliv) [reply test/hooks.ts:344]: You need to narrow the type to work with it

`options.headers` here can be either of union type element: `Headers | string[][] | Record<string, string>;`

But `set` function exists only in `Headers`, that's why I cast it here

File: source/core/Ky.ts (hunk 16 above)
Comment (Beraliv) [reply source/core/Ky.ts:156]: Removed it 🔥

File: test/hooks.ts (hunk 11 above)
Comment (Beraliv) [reply test/hooks.ts:280]: Okay, let's try it in the future if interested

**Comment by Beraliv** [comment]
@sindresorhus do you know how to solve test errors?

I have this error message:

```bash
TypeError [ERR_INVALID_RETURN_PROPERTY_VALUE]: Expected string to be returned for the "format" from the "loader getFormat" function but got type object.
```

it happens in several files:

```bash
test/retry.ts
test/browser.ts
test/main.ts
test/hooks.ts
test/headers.ts
test/prefix-url.ts
test/methods.ts
```

I use:

```bash
beraliv@Alexeys-MBP ky % node -v
v12.21.0
beraliv@Alexeys-MBP ky % npm -v
7.11.1
beraliv@Alexeys-MBP ky % yarn -v
1.22.10
```

Also have another error:

```bash
fetch.ts › fetch option takes a custom fetch function

  Planned for 12 assertions, but got 6.
```

I'm just not familiar with `ava` and how it's configured with `node`
Maybe I need something to change?

File: source/core/Ky.ts (hunk 15 above)
Comment (Beraliv) [reply source/core/Ky.ts:149]: This way it removes one `@ts-expect-error` so that's better than before

I will keep it ✅

File: test/hooks.ts (hunk 18 above)
Comment (sindresorhus) [reply test/hooks.ts]: I think `NormalizedOptions` needs to ensure headers are only `Headers`. We should present simple unified types to the user.

**Comment by sindresorhus** [comment]
> do you know how to solve test errors?

Upgrade `ts-node` to latest version.

File: source/core/Ky.ts (hunk 6 above)
Comment (lukaw3d) [reply source/core/Ky.ts:151]: Sorry to give you merge conflicts. I added `entry as [keyof typeof responseTypes, string]` in 
https://github.com/sindresorhus/ky/pull/355/commits/3e75d74d7b0eb6385d19d6ff1d080eb0f0e7834a#diff-1d28e040cea335f5a22aa9c3f678f5d00b353e3ce966dc0978a91fe56f240b04R151-L152

Comment (Beraliv) [reply source/core/Ky.ts:151]: Good, I will update the PR on the weekend, should be fine 🔥

**Comment by Beraliv** [comment]
> > do you know how to solve test errors?
> 
> Upgrade `ts-node` to latest version.

Updated 👍

**Comment by sindresorhus** [comment]
You need to fix the merge conflicts.

**Comment by Beraliv** [comment]
@sindresorhus rebased 👍

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/ky/pull/339#discussion_r626105488 was not done, so I moved it back in https://github.com/sindresorhus/ky/pull/339/commits/72de135b55a39f5088dcf9b9ea608b41a50dc695. It has to be `unknown`.

**Comment by sindresorhus** [comment]
Thanks :)

**Comment by Beraliv** [comment]
> [#339 (comment)](https://github.com/sindresorhus/ky/pull/339#discussion_r626105488) was not done, so I moved it back in [72de135](https://github.com/sindresorhus/ky/commit/72de135b55a39f5088dcf9b9ea608b41a50dc695). It has to be `unknown`.

Ah, right, thank you!

**Comment by Beraliv** [comment]
@sindresorhus thank you for the help during development, that was an interesting experience
